plugin-server/src/worker/ingestion/properties-updater.ts

```diff
@@ -66,60 +85,133 @@ export async function updatePersonProperties(
     }
 }
 
-export function calculateUpdatedProperties(
-    person: Person,
+export async function upsertGroup(
+    db: DB,
+    teamId: TeamId,
+    groupTypeIndex: GroupTypeIndex,
+    groupKey: string,
+    properties: Properties,
+    timestamp: DateTime
+): Promise<void> {
+    try {
+        const [propertiesUpdate, createdAt, version] = await db.postgresTransaction(async (client) => {
```

Comment:
nit: I think this is a bit messy. we do too much handling for both cases (insert and update) lower down with `!group` and `propertiesUpdate.updated`.

Could we do something like this (written inline):

```js
            const group: Group | undefined = await db.fetchGroup(teamId, groupTypeIndex, groupKey, client, {
                forUpdate: true,
            })
            if (!group) {
                // this will throw with a RaceConditionError if the group was already inserted
                await db.insertGroup(
                    teamId,
                    groupTypeIndex,
                    groupKey,
                    propertiesUpdate.properties,
                    timestamp,
                    propertiesUpdate.properties_last_updated_at,
                    propertiesUpdate.properties_last_operation,
                    0,
                    client
                )
                return [{...propertiesUpdate, updated: true}, timestamp, 0]
            }

            const propertiesUpdate = calculateUpdate(
                group?.group_properties || {},
                properties,
                {},
                group?.properties_last_updated_at || {},
                group?.properties_last_operation || {},
                timestamp
            )

            await db.updateGroup(
                teamId,
                groupTypeIndex,
                groupKey,
                propertiesUpdate.properties,
                group.created_at,
                propertiesUpdate.properties_last_updated_at,
                propertiesUpdate.properties_last_operation,
                group.version + 1,
                client
            )

            return [propertiesUpdate, createdAt, version]
```

At least in my opinion something like this makes the separation of the paths more clear, following a logical path of inserting if no group is found and updating otherwise.

Reply:
That doesn't work well - in the offered solution we end up not populating last_updated_at or last_operation.

If we were to move `calculateUpdate(` before `!group` check it's pretty much the same as it currently is, but slightly longer since insertGroup/updateGroup argument list is quite long.

Reply:
Didn't think my inline code would be 100% correct :D 

But yeah I see, I forgot that function doesn't just return the properties (I read it, but too much to keep in my head at once). 

Either way, it was mostly about all the handling with things like `db[upsertMethod]` and stuff. It does work though, which is what's the most important.